craft_application/services/package.py

```diff
@@ -44,6 +47,22 @@ def pack(self, prime_dir: pathlib.Path, dest: pathlib.Path) -> list[pathlib.Path
     def metadata(self) -> models.BaseMetadata:
         """The metadata model for this project."""
 
+    def update_project(self) -> None:
+        """Update project fields with dynamic values set during the lifecycle."""
+        update_vars: dict[str, str] = {}
+        project_info = self._services.lifecycle.project_info
+        for var in self._app.project_variables:
+            update_vars[var] = project_info.get_project_var(var)
+
+        emit.debug(f"Update project variables: {update_vars}")
+        self._project.__dict__.update(update_vars)
+
+        for field in self._app.mandatory_adoptable_fields:
+            if not getattr(self._project, field):
+                raise errors.PartsLifecycleError(
```

Comment:
Can we do this as a list of fields that weren't set so the user can fix them all in one go?

Reply:
Done!